Include session cleaning script

OdatNurd opened the pull request:
This includes a version of the script I use locally to remove defunct
projects from Sublime so that they're not offered in the project quick
switch panel.

In use it loads up the `Session.sublime-session` file and removes any
workspace entries that reference files that don't seem to exist any
longer. Backups are kept of the session files as well.

The script attempts to determine the appropriate data directory based
on the platform you run it on and also allows you to specify a data
directory specifically on the command line as well.

OdatNurd committed:
Include session cleaning script

This includes a version of the script I use locally to remove defunct
projects from Sublime so that they're not offered in the project quick
switch panel.

In use it loads up the `Session.sublime-session` file and removes any
workspace entries that reference files that don't seem to exist any
longer. Backups are kept of the session files as well.

The script attempts to determine the appropriate data directory based
on the platform you run it on and also allows you to specify a data
directory specifically on the command line as well.

keith-hall committed:
small docstring cleanup and clarification

keith-hall commented on session_cleaner/sublime_session_clean.py:
```diff
@@ -0,0 +1,103 @@
+    return os.path.expanduser("~/Library/Application Support/Sublime Text 3/")
+
+
+def clean_session():
```
we could follow the SRP more here (if we care about such things for this script) by taking an argument which would be the data directory, and then the `main` method could call `sublime_data_dir` and pass the value in to `clean_session`.

OdatNurd replied:
That's a good call. This originated from something that had the data directory hard coded so I just slipped that in, but that seems like a much better idea overall.

keith-hall commented on session_cleaner/sublime_session_clean.py:
```diff
@@ -0,0 +1,103 @@
+        workspaces = session["workspaces"]["recent_workspaces"]
+        present, missing = [], []
+        for f in workspaces:
+            present.append(f) if os.path.isfile(f) else missing.append(f)
```
I've not seen an `if` `else` statement used like this for determining which object to invoke the method on before, but that may just be due to lack of experience on my part. If I were to do this, I'd probably say `add_to = present if os.path.isfile(f) else missing\nadd_to.append(f)`, in the hope that it'd be more readable and DRY.

OdatNurd replied:
I'll buy that; change made.

If I was using C, I would have used a ternary here, but `(present if os.path.isfile(f) else missing).append(f)` looks weird in Python, or at least not as natural to my eyes as it would in other languages.

keith-hall commented on session_cleaner/sublime_session_clean.py:
```diff
@@ -0,0 +1,103 @@
+            present.append(f) if os.path.isfile(f) else missing.append(f)
+
+        if len(present) == len(workspaces):
+            return logging.info("No missing projects found")
```
if there are no missing projects, would it make sense to bail out here rather than create a backup of the session file and write the unchanged session back to disk?

keith-hall replied:
oh sorry, just realized it says `return logging.info`... so it already does this :+1:

OdatNurd replied:
The return on the logging call should make it leave without doing anything in that case, so it's alread doing that. Would it be more clear to have an explicit return on its own line? I tend to do this a lot but I don't actually know how idiomatic it's considered here.

keith-hall replied:
It may be that it's clear enough already, I just somehow missed it while skimming through... Wasn't expecting a log call to be returned I guess. A comment could be enough to clarify it, but yeah, I would definitely not miss a `return` on it's own line :D

keith-hall commented on session_cleaner/sublime_session_clean.py:
```diff
@@ -0,0 +1,103 @@
+                      separators=(',', ': '))
+
+        os.rename(session_file, bkp_file)
+        os.rename(tmp_file, session_file)
```
if the first rename succeeds and the second one fails for some reason, we could have problems... it'd be better to do it as an atomic operation somehow... but maybe not important as should be unlikely to fail.

OdatNurd replied:
Is there an easy way to do something like that more atomically? 

My line of thinking here is that if the first rename to make the backup fails, then the exception carries the code path away and everything is left as it is. 

Conversely if the second rename fails I'm guessing it would have to be due to problems renaming the `tmp_file` that we know should exist to a file that just existed and doesn't any more; in that case I'm not sure it's safe to bank on trying to rename the backup back the way it started originally since it might fail for the same reason?

keith-hall replied:
yeah probably it's not worth any special case handling here, I don't know of any way to do this atomically, except maybe to copy the existing file as a backup and then move the temp file over the top of it.

keith-hall commented on session_cleaner/sublime_session_clean.py:
```diff
@@ -0,0 +1,103 @@
+
+        logging.info("New session file saved")
+
+    except FileNotFoundError:
```
I wonder if it'd be better to move the exception handling to only cover the exact smallest block of code we expect could throw the exception. Otherwise, these exceptions could be raised in other places and the error might be misleading - for example, although it shouldn't happen, the `os.rename(tmp_file, session_file)` could fail with `FileNotFoundError` and although it'd be clear from looking at the code and the logs where it went wrong, it would indeed look at first glance of the file system like the original session file couldn't be found...

OdatNurd replied:
I would definitely agree with that. Realistically I think I would rather the whole thing be a little more granular instead of mostly just in a single method. That's 100% because this originated as just a few lines with some exception handling to catch obvious errors. 

That said, does it hurt readability to have a long string of try/except blocks? Or put another way, if there was a function for creating the new temporary session file, would it propagate the exception up to the caller or instead perhaps return the name of the temporary session on success and None on failure?

keith-hall replied:
I vote for return the name of the temporary session on success and `None` on failure rather than need the caller to handle exceptions and have to do something with it.

keith-hall reviewed (commented):
a few minor points which may be entirely unimportant here, but I thought perhaps worth mentioning anyway

keith-hall commented:
Actually, I just noticed that this doesn't work on Windows, due to the `recent_workspaces` paths looking like:

- `/D/test/project-name.sublime-workspace`
- `/C/Users/Keith/Documents/project-name.sublime-workspace`

and Python can't find the files and so all the entries get deleted. Probably we need to build a robust solution that can cope with UNC paths as well as local drive paths.

OdatNurd commented:
Well that's embarrassing. I tested it on both of the Windows machines at the office where one has no projects and one had a couple that are missing, but didn't test the situation with a project that still exists. The path style looked kosher to me so I never thought to actually check if Python works in the same style as e.g. cygwin does.

OdatNurd committed:
Refactor the script a little bit

This refactors the code a little bit to fix some style/design issues
that remained from the original version of the script, which was much
more crude.

In particular the command line handling is cleaned up a bit so that
the session cleaning script gets told the data directory instead of
looking it up.

There is also a slight change to the code that collates the lists of
existing and missing projects to be more Pythonic.

Some work is still needed to clean up the exception handling and to
get things working on Windows, where the paths stored in the session
are not in a format that Python can internally process.

OdatNurd committed:
More refactoring and cleanup

This refactors the remaining code to be more clear and clean with
regards to exception handling.

Here the operations to load, save and clean the session file are split
apart and handle their own errors, making it much easier to see where
things might go wrong and ensure that the proper error messages are
triggered.

This also creates an abstraction for checking if a workspace is valid
or not, since on Windows more work needs to be done on the path in
order to check, given that Sublime uses a non-native file name format
for these entries in the session.

OdatNurd committed:
Implement proper existence checking on Windows

This implements proper checking for the existence of Sublime workspace
files on the Windows platform, where Sublime stores files in a
different format than one might expect in some cases.

In the case of a UNC file name to the workspace, the file is stored as
one might expect except with POSIX separators, for example:

    //weyr/Public/TestProject/TestProject.sublime-workspace

Such a path works with the `os.path.isfile()` method as one might
expect (including when the host name is a numeric IP) although if the
host is not available the method predictably returns `False`, which
may or may not be desirable.

In contrast, file names that normally have a drive letter (either a
local drive or due to a drive mapping) are stored with the drive being
considered a path segment off of the root instead of a normal drive.
e.g.:

    /C/Users/micro/Desktop/TestProject/TestProject.sublime-workspace

In this case the existence check fails because the file name is not in
an appropriate format.

This change modifies the handling specifically for windows where the
file is not a UNC path (as determined by checking the prefix) and
rewriting the file to include a drive letter.

OdatNurd committed:
Improve command line handling using argparse

This alters the script to use the `argparse` module to parse the
command line, which allows us to more naturally specify the data
directory.

In addition, a `--dry-run` parameter is also implemented, which makes
the session cleaner stop prior to actually rewriting the session file.
This allows you to see what it would do without actually taking the
action.

As a result of these changes, the README is also updated.

OdatNurd commented:
Here's a new batch of changes to this to bring it a little more up to snuff.

The operations for loading and saving the session are split out so that they can handle their own errors, which makes the exception handling look cleaner.

Also this should hopefully work better (i.e. at all) on Windows. On that system, paths that refer to drives (either local or mapped) get rewritten enough for `isfile()` to recognize them. UNC paths seem to work just fine directly.

I also included using `argparse` to make the command line handling nicer and included a `--dry-run` argument so that you can see what would happen without actually touching the session file, since this will still blow away workspace entries that might technically exist but which are on network shares or external drives that are not currently available.
